### PR TITLE
snapshot: move cloning+packing logic in the pkg

### PR DESCRIPTION
**Pull request by fromanirh**
Up until now, the logic to perform snapshot was
inside the `ghw-snapshot` tool.
This is because the tool predated the inception
of the `snapshot` package.

However, we can now move most of the snapshot-related
logic into the `snapshot` package:
1. enables better testability - we add tests to demonstrate it!
2. enables code reuse (consumers of ghw can create snapshot
   programmatically)
3. enables a significant refactor, getting rid of a lot
   of repetitive code and introducing simpler, more declarative
   and more generic code.

Please note the refactoring intentionally does not include the
block devices cloning code; rationale for this decision is that
adaptation to new generic code isn't trivial (as for other functions),
so we delay this further simplification to a later PR.

Signed-off-by: Francesco Romani <fromani@redhat.com>

### CODE REVIEW

**Comment by fromanirh** [comment]
This PR is almost ready for review (I need to check again the tests and probably polish them a bit) but I think we should merge only after 0.7.0 is out.

**Comment by fromanirh** [comment]
This PR wants to solve https://github.com/jaypipes/ghw/issues/212

File: cmd/ghw-snapshot/main.go [hunk 1]
```diff
@@ -36,15 +33,6 @@ var (
...
 
-var (
-	createPseudofilePaths = []string{
-		"/proc/cpuinfo",
-		"/proc/meminfo",
-		"/etc/mtab",
-	}
-)
-
-// rootCmd represents the base command when called without any subcommands
```
Comment (jaypipes) [inline cmd/ghw-snapshot/main.go:47]: you can restore this docstring :)

File: pkg/snapshot/clonetree.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+package snapshot
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func CloneTreeInto(scratchDir string) error {
```
Comment (jaypipes) [inline pkg/snapshot/clonetree.go:16]: docstring please for all exported functions.

File: pkg/snapshot/clonetree.go [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+		}
+	}
+	return nil
+}
+
+type CopyFileOptions struct {
+	IsSymlink func(path string, info os.FileInfo) bool
+}
+
+func CopyFilesInto(fileSpecs []string, destDir string, opts *CopyFileOptions) error {
```
Comment (jaypipes) [inline pkg/snapshot/clonetree.go:303]: docstring

File: pkg/snapshot/clonetree.go [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+			if _, err = f.Write(buf); err != nil {
+				return err
+			}
+			f.Close()
+		}
+	}
+	return nil
+}
+
+type CopyFileOptions struct {
```
Comment (jaypipes) [inline pkg/snapshot/clonetree.go:299]: docstring

File: pkg/snapshot/clonetree.go [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+	IsSymlink func(path string, info os.FileInfo) bool
```
Comment (jaypipes) [inline pkg/snapshot/clonetree.go:300]: docstring for this attribute please. Also, perhaps we might rename this to "IsSymlinkFunction" or "IsSymlinkFn"?

Comment (jaypipes) [reply pkg/snapshot/clonetree.go:300]: you might include in the docstring an indication of usage, too.

File: pkg/snapshot/pack.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func PackFrom(snapshotName, sourceRoot string) error {
```
Comment (jaypipes) [inline pkg/snapshot/pack.go:19]: docstring

File: pkg/snapshot/pack.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	f, err := OpenDestination(snapshotName)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	return PackWithWriter(f, sourceRoot)
+}
+
+func OpenDestination(snapshotName string) (*os.File, error) {
```
Comment (jaypipes) [inline pkg/snapshot/pack.go:29]: docstring

File: pkg/snapshot/pack.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+			return nil, err
+		}
+		if fs.Size() > 0 {
+			return nil, fmt.Errorf("File %s already exists and is of size >0", snapshotName)
+		}
+	}
+	return f, nil
+}
+
+func PackWithWriter(fw io.Writer, sourceRoot string) error {
```
Comment (jaypipes) [inline pkg/snapshot/pack.go:55]: docstring

**Review by jaypipes — APPROVED** [review]
This is pretty awesome @fromanirh. I have a few little suggestions inline, but nothing big. I'd love to target this for the 0.7.0 milestone.

File: cmd/ghw-snapshot/main.go (hunk 1 above)
Comment (fromanirh) [reply cmd/ghw-snapshot/main.go:47]: sure, done

File: pkg/snapshot/clonetree.go (hunk 3 above)
Comment (fromanirh) [reply pkg/snapshot/clonetree.go:303]: done

File: pkg/snapshot/clonetree.go (hunk 2 above)
Comment (fromanirh) [reply pkg/snapshot/clonetree.go:16]: done

File: pkg/snapshot/clonetree.go (hunk 4 above)
Comment (fromanirh) [reply pkg/snapshot/clonetree.go:299]: done

File: pkg/snapshot/clonetree.go (hunk 5 above)
Comment (fromanirh) [reply pkg/snapshot/clonetree.go:300]: good idea, done.

File: pkg/snapshot/pack.go (hunk 6 above)
Comment (fromanirh) [reply pkg/snapshot/pack.go:19]: done

File: pkg/snapshot/pack.go (hunk 8 above)
Comment (fromanirh) [reply pkg/snapshot/pack.go:55]: done

**Comment by fromanirh** [comment]
I'm happy with this PR now and I addressed all the review comments.

File: pkg/snapshot/pack.go (hunk 7 above)
Comment (fromanirh) [reply pkg/snapshot/pack.go:34]: fixed

**Comment by fromanirh** [comment]
couple fixes for the CI failure and another issue spotted on Fedora 33 - which just happened to enable zram for everyone

File: README.md [hunk 9]
```diff
@@ -131,6 +131,47 @@ cpu, err := ghw.CPU(ghw.WithSnapshot(ghw.SnapshotOptions{
...
+}
+
+// optionally, you may add extra content into your snapshot.
+// ghw will ignore the extra content.
+// Glob patterns like `filepath.Glob` are supported.
+fileSpecs := []string{
+	"/proc/cmdline",
+}
+// options allows the client code to optionally deference symlinks, or copy
+// them into the cloned tree as symlinks
```
Comment (jaypipes) [inline README.md:164]: for the above comment to make sense, it probably would be good to do something like this:

```go
var opts *snapshot.CopyFileOptions
if err := snapshot.CopyFilesInto(fileSpecs, scratchDir, opts); err != nil {
```

File: pkg/snapshot/clonetree.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+	if _, err = f.Write(buf); err != nil {
+		return err
+	}
+	f.Close()
+	return nil
+}
+
+// CopyFileOptions allows to finetune the behaviour of the CopyFilesInto function
+type CopyFileOptions struct {
+	// isSymlinkFn allows to control the behaviour when handling a symlink.
```
Comment (jaypipes) [inline pkg/snapshot/clonetree.go:111]: femto-nit: should be `IsSynlinkFn`

File: pkg/snapshot/clonetree.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+	// If this hook returns true, the source file is treated as symlink: the cloned
+	// tree will thus contain a symlink, with its path adjusted to match the relative
+	// path inside the cloned tree. If return false, the symlink will be deferred.
+	// The easiest use case of this hook is if you want to avoid symlinks in your cloned
+	// tree (having duplicated content). In this case you can just add a functtion
```
Comment (jaypipes) [inline pkg/snapshot/clonetree.go:116]: s/functtion/function/

File: README.md [hunk 12]
```diff
@@ -131,6 +131,47 @@ cpu, err := ghw.CPU(ghw.WithSnapshot(ghw.SnapshotOptions{
...
+You can create ghw snapshots in two ways.
+You can just consume the `ghw-snapshot` tool, or you can create them programmatically
+from your golang code. We explore now the latter case.
+
+Snapshotting takes two phases:
+1. clone the relevant pseudofiles/pseudodirectories into a temporary tree
+   This tree is usually deleted once the packing is successful.
+2. pack the cloned tree into a tar.gz
+
+```go
```
Comment (jaypipes) [inline README.md:145]: maybe put the import statement above here?

**Review by jaypipes — COMMENTED** [review]
This is super close, @fromanirh. Couple nits inline but definitely want to get this into 0.7.0

File: README.md (hunk 9 above)
Comment (fromanirh) [reply README.md:164]: good point. Done.

File: pkg/snapshot/clonetree.go (hunk 10 above)
Comment (fromanirh) [reply pkg/snapshot/clonetree.go:111]: femto maybe, fair for sure, fixing.

File: pkg/snapshot/clonetree.go (hunk 11 above)
Comment (fromanirh) [reply pkg/snapshot/clonetree.go:116]: fixed

File: README.md (hunk 12 above)
Comment (fromanirh) [reply README.md:145]: right, done.

**Review by jaypipes — APPROVED** [review]
molto buona! :)

**Pull request merged**